mono/mini/mini-wasm-debugger.c

```diff
@@ -0,0 +1,641 @@
+
+//functions exported to be used by JS
+EMSCRIPTEN_KEEPALIVE int mono_wasm_set_breakpoint (const char *assembly_name, int method_token, int il_offset);
+EMSCRIPTEN_KEEPALIVE int mono_wasm_current_bp_id (void);
+EMSCRIPTEN_KEEPALIVE void mono_wasm_enum_frames (void);
+EMSCRIPTEN_KEEPALIVE void mono_wasm_get_var_info (int scope, int pos);
+EMSCRIPTEN_KEEPALIVE void mono_wasm_clear_all_breakpoints (void);
+EMSCRIPTEN_KEEPALIVE void mono_wasm_setup_single_step (int kind);
+
+//JS functions imported that we use
+extern void mono_wasm_add_frame (int il_offset, int method_token, const char *assembly_name);
+extern void mono_wasm_fire_bp (void);
+extern void mono_wasm_add_bool_var (gint8);
+extern void mono_wasm_add_int_var (gint32);
+extern void mono_wasm_add_long_var (gint64);
+extern void mono_wasm_add_float_var (float);
+extern void mono_wasm_add_double_var (double);
+extern void mono_wasm_add_string_var (const char*);
+
+//FIXME move all of those fields to the profiler object
+static gboolean debugger_enabled;
+
+static int event_request_id;
+#define THREAD_TO_INTERNAL(thread) (thread)->internal_thread
+
+static void
+inplace_tolower (char *c)
+{
+	int i;
+	for (i = strlen (c) - 1; i >= 0; --i)
+		c [i] = tolower (c [i]);
+}
+
+static void
+jit_done (MonoProfiler *prof, MonoMethod *method, MonoJitInfo *jinfo)
+{
+	mono_de_add_pending_breakpoints (method, jinfo);
+}
+
+/* Frame state handling */
+static GPtrArray *frames;
+
+static void
+free_frame (DbgEngineStackFrame *frame)
+{
+	g_free (frame);
+}
+
+static gboolean
+collect_frames (MonoStackFrameInfo *info, MonoContext *ctx, gpointer data)
+{
+	SeqPoint sp;
+	MonoMethod *method;
+
+	//skip wrappers
+	if (info->type != FRAME_TYPE_MANAGED && info->type != FRAME_TYPE_INTERP)
+		return FALSE;
+
+	if (info->ji)
+		method = jinfo_get_method (info->ji);
+	else
+		method = info->method;
+
+	if (!method)
+		return FALSE;
+
+	DEBUG_PRINTF (2, "Reporting method %s native_offset %d\n", method->name, info->native_offset);
+
+	if (!mono_find_prev_seq_point_for_native_offset (mono_get_root_domain (), method, info->native_offset, NULL, &sp))
+		DEBUG_PRINTF (1, "Failed to lookup sequence point\n");
+
+	DbgEngineStackFrame *frame = g_new0 (DbgEngineStackFrame, 1);
+
+	frame->ji = info->ji;
+	frame->domain = info->domain;
+	frame->method = method;
+	frame->native_offset = info->native_offset;
+
+	g_ptr_array_add (frames, frame);
+
+	return FALSE;
+}
+
+static void
+free_frame_state (void)
+{
+	if (frames) {
+		int i;
+		for (i = 0; i < frames->len; ++i)
+			free_frame (g_ptr_array_index (frames, i));
+		g_ptr_array_set_size (frames, 0);
+	}	
+}
+
+static void
+compute_frames (void) {
+	if (frames) {
+		int i;
+		for (i = 0; i < frames->len; ++i)
+			free_frame (g_ptr_array_index (frames, i));
+		g_ptr_array_set_size (frames, 0);
+	} else {
+		frames = g_ptr_array_new ();
+	}
+
+	mono_walk_stack_with_ctx (collect_frames, NULL, MONO_UNWIND_NONE, NULL);	
```

Comment:
Does this work correctly on wasm? Nice! I thought we didn’t get wasm stack walks because there’s no defined format for the runtime stack.

Reply:
We can stackwalk because the interp has a shadow stack.